### PR TITLE
[VMware] Start VM with deploy-as-is template having multiple controller types

### PR DESCRIPTION
**Pull request by sureshanaparti**
### Description
This PR fixes the deploy / start VM issue in VMware, with deploy-as-is template (OVA) having multiple controller types.

Fixes: #5111 

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [x] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?

Manually tested deploy / start using OVA template having multiple controller types.

CS-VM-with-multiple-controller-types

### CODE REVIEW

**Comment by sureshanaparti** [comment]
@blueorangutan package

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java [hunk 1]
```diff
@@ -2632,6 +2632,15 @@ private void mapSpecDisksToClonedDisksAndTearDownDatadisks(VirtualMachineMO vmMo
...
                 List<VirtualDisk> vmDisks = vmMo.getVirtualDisks();
+                Collections.sort(vmDisks, new Comparator<VirtualDisk>() {
+                    @Override
+                    public int compare(VirtualDisk disk1, VirtualDisk disk2) {
+                        Integer controllerKey1 = disk1.getControllerKey();
+                        Integer controllerKey2 = disk2.getControllerKey();
+                        return controllerKey1.compareTo(controllerKey2);
+                    }
+                });
+
```
Comment (DaanHoogland) [inline plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java:2643]: can you extract this utility?

**Review by DaanHoogland — APPROVED** [review]
code looks good, if this really solves the issue : merge.
@sureshanaparti is sorting really all that is needed? Did you try multiple orderings of disks in the template ?

**Comment by sureshanaparti** [comment]
> code looks good, if this really solves the issue : merge.
> @sureshanaparti is sorting really all that is needed? Did you try multiple orderings of disks in the template ?

@DaanHoogland the disks are returns in random order, and so there is mismatch in the mapping of disks.

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java (hunk 1 above)
Comment (sureshanaparti) [reply plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java]: > can you extract this utility?

done

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test centos7 vmware-67u3

**Review by rhtyd — APPROVED** [review]

**Comment by nvazquez** [comment]
@blueorangutan test centos7 vmware-67u3

**Review by nvazquez — APPROVED** [review]

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test centos7 vmware-67u3

**Review by nvazquez — APPROVED** [review]
LGTM - Tested using deploy-as-is on templates with multiple controller types

**Pull request merged**